### PR TITLE
Authoring Libraries

### PR DESCRIPTION
**Pull request by marioacc**
_Authoring Libraries_
A little bit of more explanation was added as well as a warning regarding to [issue 4824](https://github.com/webpack/webpack/issues/4824)

[1]: https://cla.js.foundation/webpack/webpack.js.org
[2]: https://webpack.js.org/writers-guide/

### CODE REVIEW

**Comment by skipjack** [comment]
@marioacc awesome thank you for jumping on this, I will review shortly. Can you sign the CLA? Once you've done it once, you should be all set for future contributions from the same GitHub account.

**Comment by skipjack** [comment]
@marioacc the first thing I'm noticing is the CI failure. This is due to simple markdown formatting issues:

``` bash
> webpack.js.org@0.1.2 lint:markdown /home/travis/build/webpack/webpack.js.org
> markdownlint --config ./.markdownlint.json *.md ./content/**/*.md

./content/guides/author-libraries.md: 31: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 38: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 94: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 96: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 111: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 221: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 269: MD031 Fenced code blocks should be surrounded by blank lines
./content/guides/author-libraries.md: 113: MD032 Lists should be surrounded by blank lines
./content/guides/author-libraries.md: 130: MD032 Lists should be surrounded by blank lines
./content/guides/author-libraries.md: 271: MD032 Lists should be surrounded by blank lines
```

If you could start by resolving those, that would be great.

**Comment by marioacc** [comment]
@skipjack I already have signed the CLA, so when I click in the Details link, it tells me the CLA is already signed.
I will solve the markdown issues.

**Comment by skipjack** [comment]
@marioacc the issue seems to be that one of your commits was authored with a different user (this can happen if you push from work for example). If you rebase on master with your main account, which we have to anyway to resolve the conflicts, it should fix the CLA issues.

File: content/guides/author-libraries.md [hunk 1]
```diff
@@ -6,14 +6,57 @@ contributors:
   - johnstew
   - simon04
   - 5angel
+  - marioacc
 ---
 
-webpack is a tool which can be used to bundle application code and also to bundle library code. If you are the author of a JavaScript library and are looking to streamline your bundle strategy then this document will help you.
+webpack is a tool which can be used to bundle application code and also to bundle library code. If you are the author of a JavaScript library and are looking to streamline your bundle strategy then this document will help you with the differents webpack configurations to expose your libraries as you think convenient.
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: s/differents/different

g/as you think/as you think is

File: content/guides/author-libraries.md [hunk 2]
```diff
@@ -34,56 +77,70 @@ export function wordToNum(word) {
 };
 ```
 
-The usage spec for the library will be as follows.
+The usage specification for the library use will be as follows:
 
 ```javascript
-import * as webpackNumbers from 'webpack-numbers';
-
+import * as webpackNumbers from 'webpack-numbers';//ES2015 module import
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: Nitpicking, but can you add aspace *before* the comments?
Also decide on whether or not to use a space after the `//`, as currently both seem to be used ;)

File: content/guides/author-libraries.md [hunk 3]
```diff
@@ -34,56 +77,70 @@ export function wordToNum(word) {
...
+    webpackNumbers.wordToNum('Five') 
+    /* Property in the window object */
+    window.webpackNumbers.wordToNum('Five') 
+    //
     ...
 </script>
 </html>
 ```
 
+The configurations also can expose the library in the next ways:
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: g/next/following

File: content/guides/author-libraries.md [hunk 4]
```diff
@@ -34,56 +77,70 @@ export function wordToNum(word) {
...
+
+- Property in the global object, for node.
+- Property in the this object.
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: Maybe use `this` in back ticks so it shows up as code.

File: content/guides/author-libraries.md [hunk 5]
```diff
@@ -34,56 +77,70 @@ export function wordToNum(word) {
...
+
+
 For full library configuration and code please refer to [webpack-library-example](https://github.com/kalcifer/webpack-library-example)
 
 
 ## Configure webpack
 
-Now the agenda is to bundle this library
+Now the agenda is to bundle this library achieving the next goals:
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: Not sure about the word "agenda" here, seems overly formal.
Maybe use "plan" or "idea"?

File: content/guides/author-libraries.md [hunk 6]
```diff
@@ -34,56 +77,70 @@ export function wordToNum(word) {
...
+
+- Without bundling `lodash`, but requiring it to be loaded by the consumer using `externals`.
+- Setting the library name as `webpack-numbers`.
+- Exposing the library as a variable called `webpackNumbers`.
+- Being able to access the library inside Node.js.s
+
+Also, the consumer will be able to access` the library the `next ways:
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: This sentence has a stray code block and a grammar issue.

How about: "Also, the consumer will be able to access the library in the following ways:"

File: content/guides/author-libraries.md [hunk 7]
```diff
@@ -166,11 +222,29 @@ module.exports = {
 };
 ```
 
+W>At the moment of webpack 3.5.5, using the next configuration is not working properly as stated in the [issue 4824](https://github.com/webpack/webpack/issues/4824):
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: This sentence doesn't flow nicely, how about:

"With webpack 3.5.5, using the following configuration doesn't work properly as stated in [issue 4824](https://github.com/webpack/webpack/issues/4824):"

File: content/guides/author-libraries.md [hunk 8]
```diff
@@ -166,11 +222,29 @@ module.exports = {
...
+
+W> However, you can set libraryTarget.var='_' to expect the library as a global variable
+
 ### Add `libraryTarget`
 
-For widespread use of the library, we would like it to be compatible in different environments, i. e. CommonJS, AMD, Node.js and as a global variable.
+For widespread use of the library, we would like it to be compatible in different environments, i.e. CommonJS, AMD, Node.js and as a global variable.
 
-To make your library available for reuse, add `library` property in webpack configuration.
+To make your library available for reuse, add `library` property inside `output` in webpack configuration.
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: g/`library` property/the `library` property
g/in webpack configuration/in the webpack configuration

File: content/guides/author-libraries.md [hunk 9]
```diff
@@ -185,7 +259,7 @@ module.exports = {
 };
 ```
 
-This makes your library bundle to be available as a global variable when imported. To make the library compatible with other environments, add `libraryTarget` property to the config.
+This makes your library bundle to be available as a global variable named `webpackNumbers` when imported. To make the library compatible with other environments, add `libraryTarget` property to the config. This will add the differents options about how the library can be exposed.
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: g/to be available/available
s/differents/different

File: content/guides/author-libraries.md [hunk 10]
```diff
@@ -195,13 +269,21 @@ module.exports = {
...
         library: 'webpackNumbers',
-        libraryTarget: 'umd'
+        libraryTarget: 'umd',
     }
     ...
 };
 ```
 
-If `library` is set and `libraryTarget` is not, `libraryTarget` defaults to `var` as specified in the [output configuration documentation](/configuration/output). See [`output.libraryTarget`](/configuration/output#output-librarytarget) there for a detailed list of all available options.
+You can expose the library in the next ways:
```
Comment (TheDutchCoder) [inline content/guides/author-libraries.md]: g/next/following

File: content/guides/author-libraries.md (hunk 1 above)
Comment (marioacc) [reply content/guides/author-libraries.md]: Perfect, will make the changes.

File: content/guides/author-libraries.md (hunk 5 above)
Comment (marioacc) [reply content/guides/author-libraries.md]: I think plan is way better.

File: content/guides/author-libraries.md (hunk 2 above)
Comment (marioacc) [reply content/guides/author-libraries.md]: In the other guides the comments are used in the following way:
`// Comment text`
 `var someJSCode={}`
Do you mind if I set the comments format in that way?

File: content/guides/author-libraries.md (hunk 1 above)
Comment (skipjack) [reply content/guides/author-libraries.md]: @marioacc I would actually probably simplify/rephrase this as:

> Aside from applications, webpack can also be used to bundle JavaScript libraries. The following guide is meant for library authors looking to streamline their bundling strategy.

What do you think?

File: content/guides/author-libraries.md (hunk 2 above)
Comment (skipjack) [reply content/guides/author-libraries.md]: > Also decide on whether or not to use a space after the `//`, as currently both seem to be used ;)

@TheDutchCoder yeah we should probably try to standardize it by padding both sides.

> Do you mind if I set the comments format in that way?

That sounds good to me, though I'd probably pad that `=` sign as well (e.g. `var someJSCode = {}`) 😜.

**Comment by skipjack** [comment]
@marioacc thanks for fixing the markdown issues and the CLA thing too. If you could rebase once more on master to resolve the conflicts caused by the restructuring in #1490, that would save me time 🙏. Not a big deal if not, once you make the requested changes we should be able to get this merged 👍 .

File: content/guides/author-libraries.md (hunk 1 above)
Comment (marioacc) [reply content/guides/author-libraries.md]: I like it. The simpler, the better.

**Comment by marioacc** [comment]
@skipjack Can you please take a look at the project, I am not sure If I did the rebase in a proper way.

**Comment by skipjack** [comment]
Ah, yeah something's off here. I'll pop over to you're repo and see if I can fix it. Typically rebasing shouldn't create an extra commit. When you rebase on master you'd typically have to `git push --force origin [branch-name]` to fix the branch after. It looks like you may have done the rebase correctly but then pulled and pushed instead of just force pushing.

**Comment by skipjack** [comment]
@marioacc the issue should be fixed now. All I did was... 

- Backed up your current `master` branch to a `backup` branch (which I can't push because I don't have permission on your repo to add new branches).
- Then `git reset --hard` on `master` to the most recent commit prior to your changes.
- Then `git cherry-pick`d your most recent commits.
- Then `git push -f origin master` to update your master branch.

Please double check that everything is as it should be. Like I said, I still have a `backup` locally so we can restore anything else if necessary.

**Comment by marioacc** [comment]
@skipjack Thank you very much, I am sorry for the complications my rebase caused. 
I took a look over the author-libraries.md in src/content/guides and the old one in content/guides and they are equal.
 I will delete the content folder to be in sync with the original webpack repo.

**Comment by skipjack** [comment]
@marioacc somehow things got a bit funky here again. I'll try to resolve and then review, please don't merge or push anything else until that's done.

**Comment by skipjack** [comment]
@marioacc I reviewed and the content was great. Just added one commit to clean things up a bit before I merge.

@TheDutchCoder if you wanted to review again that would be great. I'm going to merge this once the build passes just for the sake of pushing things forward as I think it's in a pretty good state. However if you catch anything, one of us can always submit a follow up PR.

**Comment by skipjack** [comment]
Ok ran the tests locally as we're still seeing network issues on travis that cause the `lint:links` / `hyperlink` script to timeout. Everything ran fine with the exception of webpack-contrib/npm-install-webpack-plugin#118 which doesn't relate to this PR. Merging...
